### PR TITLE
Initial version of ad5529r project

### PR DESCRIPTION
**Pull request by phlfm**
## PR Description

This PR implements the TB for the new ad5529r project

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [ ] New feature (change that adds new functionality)
- [x] New test (change that adds new test program and/or testbench)
- [ ] Update (change that modifies existing functionality)
- [ ] Breaking change (has dependencies in other repositories/testbenches)
- [ ] Documentation (change that adds or modifies documentation)

## PR Checklist
- [x] I have followed the code style guidelines
- [x] I have performed a self-review of changes
- [x] I have ran all testbenches affected by this PR
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe files, Copyright etc)
- [x] I have not introduced new Warnings/Errors on compilation/elaboration/simulation
- [x] I have set the verbosity level to none for the test program

### CODE REVIEW

File: library/vip/adi/spi_vip/adi_spi_vip_pkg.sv [hunk 1]
```diff
@@ -106,6 +109,8 @@ package adi_spi_vip_pkg;
               bitqueue_push_lsb(mosi_bits, mosi_bit);
             end
             mosi_mbx.put(bitqueue_to_int(mosi_bits));
+            rx_transaction_count++;
+            this.info($sformatf("[SPI VIP] MOSI Rx #%0d: received 0x%04x", rx_transaction_count, bitqueue_to_int(mosi_bits)), ADI_VERBOSITY_HIGH);
```
Comment (IstvanZsSzekely) [inline library/vip/adi/spi_vip/adi_spi_vip_pkg.sv]: The `[SPI VIP]` notation is not required, since the info method call prints the class instance's name.

File: library/vip/adi/spi_vip/adi_spi_vip_pkg.sv [hunk 2]
```diff
@@ -207,6 +222,24 @@ package adi_spi_vip_pkg;
...
+      int miso_cleared, mosi_cleared;
+      forever begin
+        vif.wait_until_reset_asserted();
+        miso_cleared = 0;
+        mosi_cleared = 0;
+        while (miso_mbx.try_get(dummy)) miso_cleared++;
+        while (mosi_mbx.try_get(dummy)) mosi_cleared++;
+        rx_transaction_count = 0;
+        tx_transaction_count = 0;
+        this.info($sformatf("[SPI VIP] Reset asserted - cleared %0d MISO, %0d MOSI items, reset transaction counters", miso_cleared, mosi_cleared), ADI_VERBOSITY_HIGH);
```
Comment (IstvanZsSzekely) [inline library/vip/adi/spi_vip/adi_spi_vip_pkg.sv]: The number of MISO and MOSI transactions cleared from the mailbox can be done in a cleaner way, with `mailbox.num()`, and then just clearing the mailbox itself.
`while (miso_mbx.num()) miso_mbx.get();`

Comment (IstvanZsSzekely) [inline library/vip/adi/spi_vip/adi_spi_vip_pkg.sv]: You may want to use the clear_tx and clear_rx functions to do the clearing and counting, as they do exactly that, and then just call the functions here.

File: library/vip/adi/spi_vip/spi_vip_if.sv [hunk 3]
```diff
@@ -190,6 +191,20 @@ interface spi_vip_if #(
       miso_oen <= #(CS_TO_MISO*1ns) val;
     endtask
 
+    virtual function logic get_resetn();
+      return resetn;
+    endfunction : get_resetn
+
+    virtual task wait_until_reset_asserted();
+      if (resetn == 1'b1)
+        @(negedge resetn);
```
Comment (IstvanZsSzekely) [inline library/vip/adi/spi_vip/spi_vip_if.sv]: Please use the begin...end blocks, even if it's just 1 line that's being affected by the if statement.

File: testbenches/project/ad5529r/cfgs/cfg_stress.tcl [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+set ad_project_params(NUM_OF_WORDS)         17
+# Stress test: number of frames (default 100, configurable via environment variable)
+# Usage: NUM_OF_TRANSFERS=500 make cfg_stress:test_program
+if {[info exists ::env(NUM_OF_TRANSFERS)]} {
+    set ad_project_params(NUM_OF_TRANSFERS) $::env(NUM_OF_TRANSFERS)
+    puts "INFO: NUM_OF_TRANSFERS set from environment: $ad_project_params(NUM_OF_TRANSFERS)"
+} else {
+    set ad_project_params(NUM_OF_TRANSFERS) 100
+    puts "INFO: NUM_OF_TRANSFERS using default: 100"
+}
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/cfgs/cfg_stress.tcl]: This part is not supported by our CI infrastructure. Environment variables are not used for configuring the block design or the simulation. They are either set directly in the configuration file or the simulation.

File: testbenches/project/ad5529r/cfgs/cfg_ad5529r.tcl
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/cfgs/cfg_ad5529r.tcl:1]: Some of these parameters are only used in simulation.
Configuration parameters present in the configuration file are mainly used for configuring the block design. These parameters are also converted and passed on to the simulation environment as well, since there are designs that need these parameters in the simulation as well.
If these parameters are only used in the simulation, I suggest moving these parameters to the test_program, randomize and constrain them in a way, in which it makes sense. What I mean is valid configurations.

File: testbenches/project/ad5529r/Makefile [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+####################################################################################
+
+# Makeincludes
+include ../../../scripts/make_tb_path.mk
+include $(ADI_TB_DIR)/library/includes/Makeinclude_dmac.mk
+include $(ADI_TB_DIR)/library/includes/Makeinclude_spi_engine.mk
+
+# Remaining test-bench dependencies except test programs
+SV_DEPS += $(ADI_TB_DIR)/library/regmaps/adi_regmap_clkgen_pkg.sv
+SV_DEPS += $(ADI_TB_DIR)/library/regmaps/adi_regmap_pwm_gen_pkg.sv
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/Makefile]: Please use their appropriate Makeinclude files instead.

File: testbenches/project/ad5529r/Makefile [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+LIB_DEPS += axi_clkgen
+LIB_DEPS += axi_pwm_gen
+LIB_DEPS += axi_dmac
+LIB_DEPS += util_axis_fifo
+LIB_DEPS += spi_engine/axi_spi_engine
+LIB_DEPS += spi_engine/spi_engine_execution
+LIB_DEPS += spi_engine/spi_engine_interconnect
+LIB_DEPS += spi_engine/spi_engine_offload
+
+SIM_LIB_DEPS += spi_vip
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/Makefile]: The SPI VIP is included in the SPI Engine include file.

File: testbenches/project/ad5529r/system_bd.tcl [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+# ***************************************************************************
+
+global ad_project_params
+
+source $ad_hdl_dir/library/spi_engine/scripts/spi_engine.tcl
+
+adi_project_files [list \
+    "$ad_hdl_dir/library/common/ad_edge_detect.v" \
+    "$ad_hdl_dir/library/util_cdc/sync_bits.v" \
+    "$ad_hdl_dir/library/common/ad_iobuf.v" \
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/system_bd.tcl]: These Verilog files are not used in the block design.

File: testbenches/project/ad5529r/system_project.tcl [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+source "cfgs/${cfg_file}"
+
+# Set the project name
+set project_name [file rootname $cfg_file]
+
+# Set to use SmartConnect or AXI Interconnect
+set use_smartconnect 1
+
+# Create the project
+adi_sim_project_xilinx $project_name "xc7z007sclg400-1"
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/system_project.tcl:20]: Please rebase the project on current main. The updated infrastructure doesn't support this variant of FPGA part setting.

File: testbenches/project/ad5529r/system_project.tcl [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+source $ad_tb_dir/library/includes/sp_include_dmac.tcl
+source $ad_tb_dir/library/includes/sp_include_spi_engine.tcl
+
+# Add test files to the project
+adi_sim_project_files [list \
+  "$ad_tb_dir/library/regmaps/adi_regmap_clkgen_pkg.sv" \
+  "$ad_tb_dir/library/regmaps/adi_regmap_pwm_gen_pkg.sv" \
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/system_project.tcl]: Please source their appropriate include files instead.

File: testbenches/project/ad5529r/system_tb.sv [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+  assign spi_sclk = `TH.`SPI_S.inst.IF.s_sclk;
+  assign spi_cs   = `TH.`SPI_S.inst.IF.s_cs;
+  assign spi_mosi = `TH.`SPI_S.inst.IF.s_mosi;
+  assign spi_miso = `TH.`SPI_S.inst.IF.s_miso;
+
+  // SPI VIP reset signal - directly driven by test_program for precise timing control
+  // Must be asserted BEFORE system reset so VIP knows to expect CS glitches
+  wire spi_resetn;
+  assign `TH.`SPI_S.inst.IF.resetn = spi_resetn;
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/system_tb.sv]: Please keep the system_tb "clean. What I mean by this is: don't have any signal assignment or processing in this file. Just connect the relevant signals between the Test harness and the Test program, or hard wire some of them to GND or VCC if needed.

File: testbenches/project/ad5529r/tests/test_program.sv [hunk 11]
```diff
@@ -0,0 +1,1839 @@
...
+endtask
+
+// --------------------------
+// Wrapper function for AXI write
+// --------------------------
+task axi_write(
+    input [31:0]  waddr,
+    input [31:0]  wdata);
+  base_env.mng.master_sequencer.RegWrite32(waddr,wdata);
+endtask
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/test_program.sv]: Please remove these functions and use the appropriate API function calls instead.

File: testbenches/project/ad5529r/tests/test_program.sv [hunk 12]
```diff
@@ -0,0 +1,1839 @@
...
+// State Reset Between Test Cases
+//---------------------------------------------------------------------------
+task reset_dut_state();
+  `INFO(("[RESET] Resetting DUT state..."), ADI_VERBOSITY_LOW);
+
+  // Disable offload first
+  axi_write(`SPI_ENGINE_SPI_REGMAP_BA + GetAddrs(AXI_SPI_ENGINE_OFFLOAD0_EN), 0);
+
+  // Wait for any pending SPI transactions to complete
+  wait_random(400, 600);
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/test_program.sv]: When calling a method, please specify the argument's name before passing the values to them. This will prevent confusion and makes the code more verbose.
```
wait_random(
  .min_ns(400),
  .max_ns(600));
```

File: testbenches/project/ad5529r/tests/test_program.sv [hunk 13]
```diff
@@ -0,0 +1,1839 @@
...
+// Wrapper function for clearing SPI send buffer
+// --------------------------
+task spi_clear_send();
+  spi_env.spi_agent.sequencer.clear_send();
+endtask
+
+// --------------------------
+// Random delay utility (deterministic with same seed)
+// --------------------------
+task wait_random(input int min_ns, input int max_ns);
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/test_program.sv]: If a method has more than one argument, please break each argument into a separate line.

File: testbenches/project/ad5529r/tests/test_program.sv [hunk 14]
```diff
@@ -0,0 +1,1839 @@
...
+endtask
+
+// Wait for system reset to complete (deassert)
+task wait_for_reset_complete();
+  if (!system_reset_complete) begin
+    `INFO(("[RESET_TEST] Waiting for system reset to complete..."), ADI_VERBOSITY_LOW);
+    @(posedge system_reset_complete);
+  end
+  // Additional settling time after reset
+  #100ns;
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/test_program.sv]: This delay is relatively small and can cause issues if the main clock is changed. Please use a larger delay if settling time is needed for the simulation, or, if there is a way to check when the system is ready, then check that, or if it doesn't matter, remove it.

File: testbenches/project/ad5529r/tests/test_program.sv [hunk 15]
```diff
@@ -0,0 +1,1839 @@
...
+    `ERROR(("[SCLK] No SCLK edges detected - cannot verify frequency"));
+    total_error_count++;
+    return;
+  end
+
+  avg_period_ns = sclk_period_sum / real'(sclk_period_count);
+  actual_freq_mhz = 1000.0 / avg_period_ns;  // ns to MHz conversion
+  deviation_pct = ((actual_freq_mhz - expected_freq_mhz) / expected_freq_mhz) * 100.0;
+
+  `INFO((""), ADI_VERBOSITY_NONE);
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/test_program.sv]: `ADI_VERBOSITY_NONE` is reserved for outstanding, must be printed all the time - type of messages. Please refer to the [Reporting](https://analogdevicesinc.github.io/testbenches/user_guide/coding_guidelines.html#f-reporting)'s F5 section.

File: testbenches/project/ad5529r/tests/test_program.sv [hunk 16]
```diff
@@ -0,0 +1,1839 @@
...
+  if (deviation_pct < 0) deviation_pct = -deviation_pct;  // abs
+
+  if (deviation_pct > tolerance_pct) begin
+    // In simulation, axi_clkgen may not produce exact hardware frequencies
+    // Report as warning, not error
+    `WARNING(("[SCLK] Frequency deviation %.2f%% exceeds tolerance %.2f%% (simulation clock may differ from hardware)", deviation_pct, tolerance_pct));
+    `INFO(("  [CHECK] SCLK frequency: Status=WARN (simulation)"), ADI_VERBOSITY_NONE);
+  end else begin
+    `INFO(("  [CHECK] SCLK frequency: Status=PASS"), ADI_VERBOSITY_NONE);
+  end
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/test_program.sv]: I would suggest adding an edge case, where if deviation is much bigger than tolerance to actually throw an error, as it would signal that there is an issue either with the simulation or the HDL.

File: testbenches/project/ad5529r/tests/test_program.sv [hunk 17]
```diff
@@ -0,0 +1,1839 @@
...
+
+  `INFO(("  Transactions measured: %0d", cs_timing_samples), ADI_VERBOSITY_NONE);
+  `INFO(("  CS setup (t5):  min=%.2f ns, max=%.2f ns (required: >= %.2f ns)",
+         cs_setup_min, cs_setup_max, min_cs_setup_ns), ADI_VERBOSITY_NONE);
+
+  if (cs_setup_min < min_cs_setup_ns) begin
+    `ERROR(("[CS] Setup time violation: %.2f ns < %.2f ns minimum", cs_setup_min, min_cs_setup_ns));
+    total_error_count++;
+    `INFO(("  [CHECK] CS setup: Status=FAIL"), ADI_VERBOSITY_NONE);
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/test_program.sv]: In my opinion this line is redundant.

File: library/vip/adi/spi_vip/spi_vip_if.sv [hunk 18]
```diff
@@ -190,6 +191,20 @@ interface spi_vip_if #(
       miso_oen <= #(CS_TO_MISO*1ns) val;
     endtask
 
+    virtual function logic get_resetn();
+      return resetn;
```
Comment (IstvanZsSzekely) [inline library/vip/adi/spi_vip/spi_vip_if.sv]: There are no signals or functions that allow the control of the `resetn` signal. Please add a function that allows the signal to be controlled through simulation, or create an external signal which allows the control through the block design.
Controlling through the simulation will mean that each IP that needs to be reset will be reset manually, which may lead to resetting unwanted IPs or not resetting them all.
Controlling through the block design will allow the reset of multiple IPs with a single reset assertions, which is less prone to errors.

File: testbenches/project/ad5529r/tests/test_program.sv [hunk 19]
```diff
@@ -0,0 +1,1839 @@
...
+
+  mng = new("", `TH.`MNG_AXI.inst.IF);
+  ddr = new("", `TH.`DDR_AXI.inst.IF);
+
+  `LINK(mng, base_env, mng)
+  `LINK(ddr, base_env, ddr)
+
+  spi_env = new("SPI Environment", `TH.`SPI_S.inst.IF.vif);
+
+  setLoggerVerbosity(ADI_VERBOSITY_NONE);
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/test_program.sv]: Please set the logger verbosity level before retrieving the process's randomization state.

File: testbenches/project/ad5529r/tests/test_program.sv [hunk 20]
```diff
@@ -0,0 +1,1839 @@
...
+  `INFO(("=== AD5529R TEST SUMMARY ==="), ADI_VERBOSITY_NONE);
+  `INFO(("  Total errors: %0d", total_error_count), ADI_VERBOSITY_NONE);
+
+  if (total_error_count == 0) begin
+    `INFO(("=== AD5529R Test PASSED ==="), ADI_VERBOSITY_NONE);
+  end else begin
+    `ERROR(("=== AD5529R Test FAILED ==="));
+    `FATAL(("Test terminated with %0d errors", total_error_count));
+  end
+  $finish();
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/test_program.sv]: Please use the generic Testbench done print at the end of simulation.
```
`INFO(("Test bench done!"), ADI_VERBOSITY_NONE);
$finish();
```

**Review by IstvanZsSzekely — CHANGES_REQUESTED** [review]

**Comment by phlfm** [comment]
@IstvanZsSzekely I've fixed all the issues you pointed out and heavily refactored the TB to unify the flow for the different testcases and to facilitate future maintenance. Also rebased to the latest main.

File: testbenches/project/ad5529r/system_project.tcl (hunk 8 above)
Comment (IstvanZsSzekely) [reply testbenches/project/ad5529r/system_project.tcl:20]: The `"xc7z007sclg400-1"` part must be removed when creating the project. The board/FPGA part is either specified in the configuration file, or if it isn't specified, then a random board is chosen to do the simulation on. Since for the AD5529R project we support certain FPGA carrier board, please specify it/them in the configuration file.

File: testbenches/project/ad5529r/system_bd.tcl [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+create_bd_port -dir O ad5529r_tg1
+create_bd_port -dir O ad5529r_tg2
+create_bd_port -dir O ad5529r_tg3
+
+ad_connect ad5529r_spi_clk axi_ad5529r_clkgen/clk_0
+ad_connect ad5529r_spi_irq spi_ad5529r/irq
+ad_connect ad5529r_tg0 tg0
+ad_connect ad5529r_tg1 tg1
+ad_connect ad5529r_tg2 tg2
+ad_connect ad5529r_tg3 tg3
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/system_bd.tcl:69]: tg* are already being output, and the ad5539r_tg* just duplicates those signals, same direction and all.

File: testbenches/project/ad5529r/tests/ad5529r_test_flow.svh [hunk 22]
```diff
@@ -0,0 +1,1086 @@
...
+import `PKGIFY(test_harness, mng_axi_vip)::*;
+import `PKGIFY(test_harness, ddr_axi_vip)::*;
+
+// Shared test flow body. Not standalone: `included by a `program wrapper
+// (tests/test_*.sv) supplying the ports + `NUM_OF_WORDS / `NUM_OF_TRANSFERS.
+
+wire spi_sclk = `TH.`SPI_S.inst.IF.s_sclk;
+wire spi_cs   = `TH.`SPI_S.inst.IF.s_cs;
+wire spi_mosi = `TH.`SPI_S.inst.IF.s_mosi;
+wire spi_miso = `TH.`SPI_S.inst.IF.s_miso;
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/ad5529r_test_flow.svh:63]: The current coding style hard-wires the SPI signals to the one VIP whose name is specified. This cannot be reused anywhere else because of this.
Instead of probing the signals directly on the SPI VIP itself, I would recommend creating a method inside the VIP that allows you to probe the clock and the CS signals to return their values when they change or on request. This extends the usability of the signals to be used inside classes as well through the VIP.

File: testbenches/project/ad5529r/tests/ad5529r_test_flow.svh [hunk 23]
```diff
@@ -0,0 +1,1086 @@
...
+  fork : wait_or_timeout
+    begin : wait_count
+      wait (tput_meter.total_transfers >= `NUM_OF_TRANSFERS);
+    end
+    begin : watchdog
+      #(timeout_ns * 1ns);
+      timed_out = 1;
+    end
+  join_any
+  disable wait_or_timeout;
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/ad5529r_test_flow.svh:901]: Please create a fork "guard" around the fork that's about to be disabled. This prevents accidental fork kills, as `disable wait_or_timeout` disables every fork named wait_or_timeout. Please also change the named disable to simply fork, as it's better in terms of scoping. Note: I might be wrong on this one with the scope, it's been a while since I looked this up.
```
fork
  <main fork>
  disable fork;
join
```

File: testbenches/project/ad5529r/tests/ad5529r_test_flow.svh [hunk 24]
```diff
@@ -0,0 +1,1086 @@
...
+throughput_meter tput_meter;
+
+// PWM timing measurement
+time tg_rise_times[NUM_TG][$];
+time tg_fall_times[NUM_TG][$];
+bit pwm_measurement_enabled = 0;
+
+// Main procedure
+initial begin : main
+  setLoggerVerbosity(ADI_VERBOSITY_LOW);
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/ad5529r_test_flow.svh:292]: Please set logger verbosity to none.

File: testbenches/project/ad5529r/tests/ad5529r_test_flow.svh [hunk 25]
```diff
@@ -0,0 +1,1086 @@
...
+
+  init_environment();
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/ad5529r_test_flow.svh:294]: It is a bit weird that the base and SPI environments are started inside the init function, where they are also created, but the stop functions are called plainly in the main initial block.
Thinking about other testbenches in the meantime, it would be useful to have a function that instantiates all the environments, classes and sets up their initial relation and hierarchy. This would reduce overhead in some testbenches, which require many classes to begin with, and it also simplifies the main initial block. Having the start, stop, reset functions hidden can be a bit of an issue I suppose, as they are the most important parts of the testbench.
I'm open for suggestions and other points of view regarding this.

File: testbenches/project/ad5529r/tests/ad5529r_test_flow.svh [hunk 26]
```diff
@@ -0,0 +1,1086 @@
...
+//   2. generate data (generate_sdo_data + write_data_to_ddr)
+//   3. SPI transfers (config_tx_dma + config_offload_command_fifo + run/verify)
+//   4. verify
+task automatic run_offload_test(
+  input offload_test_t data_mode
+);
+  int num_words   = (`NUM_OF_TRANSFERS) * (`NUM_OF_WORDS);
+  int total_bytes = (num_words * `DATA_DLENGTH) / 8;
+
+  print_test_header(data_mode, total_bytes);
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/ad5529r_test_flow.svh:328]: When calling a method, please specify the argument's name before passing the values to them.
If a method has more than one argument, please break each argument into a separate line.

File: testbenches/project/ad5529r/tests/ad5529r_test_flow.svh [hunk 27]
```diff
@@ -0,0 +1,1086 @@
...
+import test_harness_env_pkg::*;
+import adi_axi_agent_pkg::*;
+import m_axi_sequencer_pkg::*;
+import axi_vip_pkg::*;
+import axi4stream_vip_pkg::*;
+import adi_regmap_pkg::*;
+import adi_regmap_clkgen_pkg::*;
+import adi_regmap_dmac_pkg::*;
+import adi_regmap_pwm_gen_pkg::*;
+import adi_regmap_spi_engine_pkg::*;
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/ad5529r_test_flow.svh:49]: Redundant import functions: `adi_regmap_*`

File: testbenches/project/ad5529r/tests/ad5529r_test_flow.svh [hunk 28]
```diff
@@ -0,0 +1,1086 @@
...
+  reset_pwm_measurement();
+  // Reset toggle_gen PWM
+  mSeq.RegWrite32 (`SPI_ENGINE_TOGGLE_GEN_BA + GetAddrs(AXI_PWM_GEN_REG_RSTN), `SET_AXI_PWM_GEN_REG_RSTN_RESET(1));
+  // Per-channel period/width (one register each, +4 bytes apart)
+  for (int ch = 0; ch < NUM_TG; ch++) begin
+    mSeq.RegWrite32 (`SPI_ENGINE_TOGGLE_GEN_BA + GetAddrs(AXI_PWM_GEN_REG_PULSE_X_PERIOD) + ch*4, PWM_PERIOD_C);
+    mSeq.RegWrite32 (`SPI_ENGINE_TOGGLE_GEN_BA + GetAddrs(AXI_PWM_GEN_REG_PULSE_X_WIDTH)  + ch*4, PWM_WIDTH_C);
+  end
+  // Load configuration
+  mSeq.RegWrite32 (`SPI_ENGINE_TOGGLE_GEN_BA + GetAddrs(AXI_PWM_GEN_REG_RSTN), `SET_AXI_PWM_GEN_REG_RSTN_LOAD_CONFIG(1));
```
Comment (IstvanZsSzekely) [inline testbenches/project/ad5529r/tests/ad5529r_test_flow.svh:1026]: Please use the PWM API for reading/writing from/to the PWM IP.
Same goes for the Clock Gen and DMA IPs.

**Review by IstvanZsSzekely — CHANGES_REQUESTED** [review]
I like this addition to be able to specify the simulation seed when running make.
I request that you extract these changes with the simulation seed setting and open a new PR for this. And please update the [documentation page](https://analogdevicesinc.github.io/testbenches/user_guide/debug_tb.html#replicating-a-simulation) as well.
Regarding the SystemVerilog Header file to have all the code inside a file and import it in one go is really interesting. I've got mixed feelings about it. The test programs look weird, but it also solves the issue of copy paste redundant code.
